Add lowerings to special functions

pearu opened the pull request:
As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119558
* __->__ #119187



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

pearu committed:
Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1.

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -511,6 +511,139 @@ def remainder(a, b):
+
+        def pointwise_factory(impl):
+            def func(x):
+                return f"{impl}({x})"
```
This won't work for `CppVecOverrides` since most operations are methods on the `Vectorized` type. It's also called pointwise but is really only unary pointwise.

pearu replied:
Fixed via introducing `cppvec` item to `pointwise_overrides_data`.

pearu commented on torch/_inductor/codegen/common.py:
```diff
@@ -511,6 +511,139 @@ def remainder(a, b):
+pointwise_overrides_data: Dict[str, Dict[str, str]] = dict(
+    airy_ai=dict(
+        cpp="airy_ai_forward",
+        # triton=None,  # libdevice/triton do not provide Airy Ai function
```
yes, the logic is not there yet. Functions that don't have libdevice/triton equivalent, can either be implemented as decompositions (e.g. `i0e`, `i1e`, etc but with a risk of having lower accuracy) or as fallbacks when on CUDA device. Currently, all such functions are fallbacks (including CPU cases). The plan is to implement fallback only when triton/cpp equivalent is not available, otherwise use the implementation that described in `pointwise_overrides_data`.

pearu replied:
> Actually no need to check, as the tests show it's falling back on CPU

Yes, however, it would be preferred if there would be no fallback on CPU.

pearu replied:
The fallback logic is now implemented via `triton_fallback` kw argument to `make_pointwise`.

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -511,6 +511,185 @@ def remainder(a, b):
+        # cppvec=...,  # not impl in aten/.../vec
+        aten="special_log_ndtr",
+    ),
+    # log_softmax
+    # logit
+    # logsumexp
```
`log_softmax` and `logsumexp` aren't pointwise operators.

pearu replied:
Right, the references to these are now removed.

peterbell10 commented on torch/_inductor/codegen/common.py:
```diff
@@ -511,6 +511,185 @@ def remainder(a, b):
+                setattr(cls, funcname, staticmethod(factory(impl)))
+
+
+pointwise_overrides_data: Dict[str, Dict[str, str]] = dict(
```
This is good as-is, but I think we could expand this to non-numeric operators as well if you changed `Dict[str, str]` to a `dataclass` and had `type_promotion_kind` be a member.

pearu replied:
Are you suggesting introducing a new dataclass that holds `cpp`, `triton`, `cppvec` members instead of using `dict`? If so, how adding another member `type_promotion_kind` will relate to supporting `non-numeric` operators?

peterbell10 replied:
Instead of calling `register_pointwise_numeric` in `lowering.py`, you could use `register_pointwise` and pass the `type_promotion_kind`. This would allow registering operators that don't promote int to float.

pearu replied:
Ok, gotcha you.

pearu replied:
I introduced `type_promotion_kind` item and switched to using `Dict[str, Any]` for now.

peterbell10 replied:
`dataclass` would be better so we could get strong typing. There are ongoing efforts to make mypy typing more strict throughout inductor.

pearu replied:
ok, I'll give it a shot.

pearu replied:
done

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -8470,23 +8474,87 @@ def return_true(*args, **kwargs):
+            dtype = torch.float64
+
+        if name in {"gammainc", "gammaincc"}:
+            args = (
```
Not 100% convinced this is worth doing for operators that have different argument formats, but oh well.

pearu replied:
At least now these are covered as well.

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -8470,23 +8474,87 @@ def return_true(*args, **kwargs):
+            "zeta",
+        }:
+            # <func>_cuda not implemented for Half
+            dtype = torch.float64
```
You can use `check_lowp=False` in `self.common` to skip the half-precision check.

peterbell10 reviewed (approved):
Seems like a nice cleanup

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to bessel_j0/j1/y0/y1, erfcx, i0, i1."

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu commented:
@pytorchbot merge

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu commented:
@pytorchbot merge

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu committed:
Update on "Add lowerings to special functions"

As in the title.

In addition, the PR introduces infrastructure for lowerings of pointwise functions that have both cpp and triton implementations available.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

pearu commented:
@pytorchbot merge
